### PR TITLE
bpo-31861: Add operator.aiter and operator.anext

### PR DESCRIPTION
**Pull request by jab**
...along with passing tests. (I added the tests to `test_asyncgen.py` rather than to `test_operator.py` because they fit much better there.)

/cc @1st1 @njsmith @JelleZijlstra et al.

https://bugs.python.org/issue31861

### CODE REVIEW

File: Lib/operator.py
Comment (asvetlov) [inline Lib/operator.py]: `__aiter__` is a regular function. Why is `aiter` async?

Comment (asvetlov) [reply Lib/operator.py]: Isn't better to avoid `*args`?
Something like
```
_sentinel = object()
def aiter(async_iterable, sentinel=_sentinel): ...
```

File: Lib/operator.py
Comment (asvetlov) [inline Lib/operator.py]: A question about `*args` again

File: Lib/operator.py
Comment (asvetlov) [inline Lib/operator.py]: `an AsyncIterator` here

**Review by asvetlov — CHANGES_REQUESTED** [review]

File: Lib/operator.py
Comment (jab) [reply Lib/operator.py]: Done in 908227b.

File: Lib/operator.py
Comment (jab) [reply Lib/operator.py]: Done in 908227b.

File: Lib/operator.py
Comment (jab) [reply Lib/operator.py]: Thanks for catching, done in 908227b.

**Comment by jab** [comment]
@asvetlov Thank you for the helpful review. I have addressed all your comments in 908227b. GitHub reports that all CI checks are passing. In particular, I see "`test_asyncgen passed`" (which contains the tests for these changes) in e.g. [this job](https://travis-ci.org/python/cpython/jobs/425818359). The latest coverage report for these changes is [here](https://codecov.io/gh/python/cpython/commit/908227bbdd15de195bed4a42fd016a99ec73dbf6).

Locally I get:
```
$ ./python.exe Lib/test/test_asyncgen.py
...........................Task was destroyed but it is pending!
task: <Task pending name='Task-45' coro=<<async_generator_athrow without __name__>()>>
......................
----------------------------------------------------------------------
Ran 49 tests in 1.185s

OK
```
I'm not sure the destroyed pending task there is my fault or whether this is even a problem.

Any further reworking needed before this can be merged?

**Comment by jab** [comment]
I have made the requested changes; please review again.

File: Lib/test/test_asyncgen.py
Comment (jab) [inline Lib/test/test_asyncgen.py:406]: I based this off the `class Gen` defined in the existing `test_async_gen_asyncio_03()` test. But based on your previous comment, should these both use `def` rather than `async def`?

**Review by 1st1 — CHANGES_REQUESTED** [review]
Please don't merge this yet. I'm not convinced that aiter and anext shouldn't be builtins.

**Comment by jab** [comment]
@1st1 Would it be useful if I submitted another change here that moved these two Python functions from `operator` into `builtins`, assuming that's possible? Or do you want more time to think about this? If so, can you relabel this back to awaiting review? (@bedevere-bot automatically changed the label here to "awaiting changes" after you requested changes.)

**Comment by jab** [comment]
Regardless, I'd be curious if you can spot anything in the tests I added that could be responsible for the lingering pending task and if there's anything I should do about that. The [question](https://github.com/python/cpython/pull/8895#discussion_r216027031) about whether the `async def __aiter__` should have been `def __aiter__` in the (existing) tests is also still outstanding, in case you can help with the answer.

**Comment by 1st1** [comment]
@jab I know that @WolkoDav was (is?) working on implementing builtin versions of these functions (in C), but I'm not sure what's his latest progress. Perhaps you guys can collaborate and work together on this.

In my opinion it would be great to have `aiter()` and `anext()` as builtin functions and I'm 80% confident that that's what we are going to do in 3.8.  The problem is that adding new builtins is a serious decision, and @gvanrossum is no longer available to grant us a permission to do that right now.  So we'll have to wait until we have a new governance model in CPython to have a final decision on this.

File: Lib/test/test_asyncgen.py
Comment (1st1) [reply Lib/test/test_asyncgen.py:406]: 1. `__aiter__` must be **a synchronous function** (`def`) and must return *an asynchronous iterator*.

2. Whan `__aiter__` is declared as an asynchronous generator, it is still **a synchronous function**, that just returns an instance of *an asynchronous generator* when you call it (so (1) still stands).

Comment (jab) [reply Lib/test/test_asyncgen.py:406]: Thanks for explaining, but I'm still confused. After reading (1) it sounded like `__aiter__` should be rewritten, but after reading (2) it sounded like the code should stay written as-is (even though it seems more confusing this way).

**Comment by 1st1** [comment]
> Rather than blocking all progress on this, would it make sense to merge this PR so we can make incremental progress in the meantime, and then once the governance question has been settled, the decision to move these into builtins could be made independently?

I don't see how merging either PR right now (with a non-zero chance of it being reverted later) can be beneficial.  I don't think this feature would see a lot of work and refinement after we merge it.  I would also like to avoid any confusing churn in APIs/docs if possible.

> (Are there any other core developers you think would want to weigh in here who we should ping?)

I think that we'll simply email python-dev and ask when the time is right.  Please be patient!

If you want to work on something else meanwhile I'm sure that there are some open issues in asyncio that could be easier to merge/fix.

File: Lib/test/test_asyncgen.py
Comment (1st1) [reply Lib/test/test_asyncgen.py:406]: It's OK for it to stay written as is.

> After reading (1) it sounded like __aiter__ should be rewritten, but after reading (2) it sounded like the code should stay written as-is (even though it seems more confusing this way).

(2) is an edge case of (1); the case of `__iter__` returning an iterator and `__iter__` being a generator function is absolutely the same.

**Comment by jab** [comment]
Ok, just trying to maximize results while I had the bandwidth now. I don't subscribe to python-dev so please link to the thread here once it's started if possible.

I actually already have another PR outstanding that's been awaiting review for a few weeks, in case it's one you feel comfortable reviewing: #8792

**Comment by WolkoDav** [comment]
> @jab I know that @WolkoDav was (is?) working on implementing builtin versions of these functions (in C), but I'm not sure what's his latest progress. Perhaps you guys can collaborate and work together on this.
> 
> In my opinion it would be great to have `aiter()` and `anext()` as builtin functions and I'm 80% confident that that's what we are going to do in 3.8. The problem is that adding new builtins is a serious decision, and @gvanrossum is no longer available to grant us a permission to do that right now. So we'll have to wait until we have a new governance model in CPython to have a final decision on this.

Yes, I work on C implementation in builtins for `aiter` and `anext`. I had problems with my laptop. I apologize for the delay.

**Comment by jab** [comment]
Hey @WolkoDav, sorry to hear you had laptop trouble and great to hear you're working on this! If/when you're ready to share your work, please link to it from here.

**Comment by WolkoDav** [comment]
@jab, i have some problem at the moment with default parameter. I need some time to understand how work coroutine at C level... At the moment you can view work in my fork: https://github.com/WolkoDav/cpython

**Comment by jab** [comment]
Thanks for sharing the link @WolkoDav, and sorry I haven't had time or enough recent C experience to help. I defer to @1st1, @asvetlov, or any other maintainers who can jump in here.

**Comment by WolkoDav** [comment]
@jab, Yuri has already watched my code and left comments there. I think that in a week I will be able to add the  parameter "default".

**Comment by nanjekyejoannah** [comment]
@ikrivosheev are you still working on this? I want to work on another PR for a  C implementation for this. Can i pick up. Am fine if you still want to work on this.

**Comment by jab** [comment]
Ping again, in case this is a good time for core devs to unstall this :)

**Comment by jab** [comment]
> I think that we'll simply email python-dev and ask when the time is right. Please be patient!

@1st1, would now (9 months later) be an ok time to ask about this?

**Review by eric-wieser — APPROVED** [review]

**Comment by jab** [comment]
Thanks for reviewing and approving this, @eric-wieser, great to see some activity here again!

Is the question of whether these should be in operator vs. builtins is still unresolved, though? If so, @1st1, do you still want (me?) to start a thread on python-dev about this when the time is right? Happy to do that now if that'd be helpful.

It'd be great for `aiter` and `anext` to be available _somewhere_ in the standard library, whether in operator or builtins, before too much longer. I'd love to submit an alternative PR that provided the builtins version, in case that ends up being the preferred choice, but would like to request a mentor for that if possible, since I still don't know my way around the C parts of CPython very well yet.

Would be sweet to get a fix for BPO-31861 landed by the time this PR turns 2 years old at end of August, and just want to help in any way I can :)

**Comment by eric-wieser** [comment]
> I want to work on another PR for a C implementation for this.

A C implementation might make more sense, especially if it could live next to the existing `callable_iterator` type:

https://github.com/python/cpython/blob/c3dd7e45cc5d36bbe2295c2840faabb5c75d83e4/Objects/iterobject.c#L174-L288

From what I can tell, that implementation of `aiter(callable, sentinel)` is pickleable, whereas the one in this PR is not.

**Review by nanjekyejoannah — COMMENTED** [review]
I think I agree, this is good feedback for @ikrivosheev but anyone can pick up these suggestions in a different PR unless @ikrivosheev commits again.

**Comment by jab** [comment]
Thanks for looking at this, @eric-wieser and @nanjekyejoannah, good to cross paths with you again here and hope you’ve been well!

I’d love to work on a C implementation in another PR (which could maybe even reuse the tests I wrote in this PR), but would need a little help getting started on the C side (I know C pretty well from a past life, but have only written 100 lines in the last 15 years) and don’t know the CPython code well yet. Would anyone with more experience be willing to mentor me for this?

File: Lib/operator.py
Comment (sweeneyde) [inline Lib/operator.py]:
```suggestion
        ait = obj.__aiter__()
        if not hasattr(ait, '__anext__'):
            raise TypeError(f"aiter() returned non-AsyncIterator of type {type(ait)!r}")
        return ait
```
I think it should be this way for a few reasons:

1. It eliminates the two separate non-raising cases and return statements/types
2. It eliminates the overhead of the extra async generator expression and is more easily implemented in C.
4. For any AsyncIterator `a`, we "should" have `a.__aiter__() is a`, so the behavior shouldn't change much.
5. For the cases where that isn't true, in analogy with the builtin `iter()`, one might expect `aiter(x)` to always call `x.__aiter__()`:

```python
from collections.abc import Iterator, AsyncIterator

class A:
    def __next__(self):
        return 1
    def __iter__(self):
        return iter([])

# "a" is already an iterator, but a.__iter__ is called anyway.
a = A()
assert isinstance(a, Iterator)
assert iter(a) is not a
assert type(iter(a)) is not type(a)
assert type(iter(a)) is type(iter([]))

############################

async def another_aiter():
    yield 1

class B:
    def __anext__(self):
        return 1
    def __aiter__(self):
        return another_aiter()

b = B()
assert isinstance(b, AsyncIterator)
# These three assertions fail, but pass with the change.
assert aiter(b) is not b
assert type(aiter(b)) is not type(b)
assert type(aiter(b)) is type(aiter(another_aiter()))
```

See the implementation of `iter()` [here](https://github.com/python/cpython/blob/master/Objects/abstract.c#L2616), which uses `PyObject_GetIter()` from [here](https://github.com/python/cpython/blob/master/Python/bltinmodule.c#L1511).
I wonder if a `PyObject_GetAsyncIter` needs to be added to the C-API so that `aiter()` and [the GET_AITER opcode](https://github.com/python/cpython/blob/master/Python/ceval.c#L2196) can use the same code.

Comment (JelleZijlstra) [reply Lib/operator.py]: Shouldn't we call `type(obj).__aiter__(obj)` (and the analogue for `__anext__`)? That's how dunder methods are generally called by the interpreter.

Comment (jab) [reply Lib/operator.py]: Thanks for reviewing! Updated this patch based on the good suggestions.

(Note, I used `isinstance(ait, AsyncIterator)` rather than the `hasattr(ait, '__anext__')` that was suggested, figuring that's the preferred spelling. Also, after rebasing on latest master, I had to move the `from collections.abc ...` imports from module scope to function scope as a workaround to avoid a circular dependency between `operator` and `collections`, which wasn't an issue when I first submitted this PR. Sounds like this PR will now serve as a reference implementation for a reimplementation in C (submitted in a separate PR) anyway, where I guess such a workaround won't be necessary?)

**Comment by jab** [comment]
Please see #23847 for a C implementation of aiter and anext added to builtins.

**Comment by gvanrossum** [comment]
Closing in favor of #23847 (adding them to builtins).

**Pull request closed without merging**
